cyipopt/scipy_interface.py

```diff
@@ -185,12 +216,46 @@ def get_bounds(bounds):
         return lb, ub
 
 
+def get_sparse_jacobian_structure(constraints, x0):
+    con_jac_is_sparse = []
+    jacobians = []
+    x0 = np.asarray(x0)
+    if isinstance(constraints, dict):
+        constraints = (constraints, )
+    if len(constraints) == 0:
+        return [], [], []
+    for con in constraints:
+        con_jac = con.get('jac', False)
+        if con_jac:
+            if isinstance(con_jac, bool):
+                _, jac_val = con['fun'](x0, *con.get('args', []))
+            else:
+                jac_val = con_jac(x0, *con.get('args', []))
+            # check if dense or sparse
+            if isinstance(jac_val, coo_array):
+                jacobians.append(jac_val)
+                con_jac_is_sparse.append(True)
+            else:
+                # Creating the coo_array from jac_val would yield to
+                # wrong dimensions if some values in jac_val are zero,
+                # so we assume all values in jac_val are nonzero
+                jacobians.append(coo_array(np.ones_like(np.atleast_2d(jac_val))))
+                con_jac_is_sparse.append(False)
+        else:
+            # we approximate this jacobian later (=dense)
+            con_val = np.atleast_1d(con['fun'](x0, *con.get('args', [])))
+            jacobians.append(coo_array(np.ones((con_val.size, x0.size))))
+            con_jac_is_sparse.append(False)
+    J = scipy.sparse.vstack(jacobians)
+    return con_jac_is_sparse, J.row, J.col
+
+
 def get_constraint_dimensions(constraints, x0):
     con_dims = []
     if isinstance(constraints, dict):
         constraints = (constraints, )
     for con in constraints:
-        if con.get('jac', False) is True:
+        if con.get('jac', False) == True:
```

Comment:
```suggestion
        if con.get('jac', False):
```

Reply:
@moorepants Your proposed change is not equivalent to `if con.get('jac', False) == True`. If `con['jac']` is a function or a lambda, `if con.get('jac', False):` will be unintentionally true. However, it should be true if and only if `con['jac']` is a `bool` and `True`. So I'd suggest reversing it again.

PS: I noticed that some of the tests are (correctly) failing on my local machine after this commit. Any idea why this isn't the case for the GitHub workflows? I double-checked it and the tests should definitely fail...

Reply:
The booleans in Python are built-in constants. Using an identity check (`is`) with `True` or `False` is idiomatic (it's more explicit and it's faster) as it checks object IDs rather than calling an object's `__eq__` method, which is what happens during an equality check (`==`). As you've said it should only be true iff `con['jac']` is `True` then it should be reverted to its state prior to this PR of `if con.get('jac', False) is True:`.

Reply:
Sorry for leading you astray.